api/client/gro_client_test.py

```diff
@@ -1,92 +1,373 @@
-            'convType': 1
+mock_data_points = [
+    {
+        'start_date': '2017-01-01T00:00:00.000Z',
+        'end_date': '2017-12-31T00:00:00.000Z',
+        'value': 40891, 'unit_id': 14,
+        'reporting_date': None,
+        'metric_id': 860032, 'item_id': 274, 'region_id': 1215,
+        'partner_region_id': 0, 'frequency_id': 9, 'source_id': 2,
+        'belongs_to': {
+            'metric_id': 860032,
+            'item_id': 274,
+            'region_id': 1215,
+            'frequency_id': 9,
+            'source_id': 2
         }
-    elif entity_type=='units' and entity_id==43:
-        return {
-            'id': 43,
-            'name': 'US Dollar (constant 2010)',
-            'baseConvFactor': { 'factor': None },
-            'convType': 0
+    }, {
+        'start_date': '2017-01-01T00:00:00.000Z',
+        'end_date': '2017-12-31T00:00:00.000Z',
+        'value': 56789, 'unit_id': 10,
+        'reporting_date': None,
+        'metric_id': 860032, 'item_id': 274, 'region_id': 1216,
+        'partner_region_id': 0, 'frequency_id': 9, 'source_id': 2,
+        'belongs_to': {
+            'metric_id': 860032,
+            'item_id': 274,
+            'region_id': 1216,
+            'frequency_id': 9,
+            'source_id': 2
         }
+    }
+]
+
+
+def mock_get_available(access_token, api_host, entity_type):
+    return list(mock_entities[entity_type].values())
+
+
+def mock_list_available(access_token, api_host, selected_entities):
+    return [dict(data_series) for data_series in mock_data_series]
+
+
+def mock_lookup(access_token, api_host, entity_type, entity_ids):
+    if isinstance(entity_ids, int):
+        # Raises a KeyError if the requested entity hasn't been mocked:
+        return mock_entities[entity_type][entity_ids]
     else:
-        raise Exception('{} {} not mocked'.format(entity_type, entity_id))
+        return {str(entity_id): mock_entities[entity_type][entity_id] for entity_id in entity_ids}
 
-client.lookup = MagicMock(
-    side_effect=mock_test_units
-)
 
-def test_convert_unit():
-    assert client.convert_unit({ 'value': 1, 'unit_id': 10 }, 10) == {
-        'value': 1,
-        'unit_id': 10
-    }
-    assert client.convert_unit({ 'value': 1, 'unit_id': 10 }, 14) == {
-        'value': 0.001,
-        'unit_id': 14
-    }
-    assert client.convert_unit({ 'value': 3, 'unit_id': 36 }, 37) == {
-        'value': 42,
-        'unit_id': 37
-    }
-    assert client.convert_unit({ 'value': 1, 'unit_id': 37 }, 36) == {
-        'value': -17.5,
-        'unit_id': 36
-    }
+def mock_get_allowed_units(access_token, api_host, metric_id, item_id):
+    return [unit['id'] for unit in mock_entities['units'].values()]
 
-    assert client.convert_unit({}, 36) == {}
 
-    with pytest.raises(Exception):
-        assert client.convert_unit({ 'value': 1, 'unit_id': 10 }, 43)
+def mock_get_data_series(access_token, api_host, **selection):
+    return [dict(data_series) for data_series in mock_data_series]
```

Comment:
Using `dict(a)` gives a copy of `a` instead of `a` itself. In `find_data_series` it calls `get_data_series()` internally and then modifies the output. Before this `dict()` change, that ended up modifying the test data which affected other tests.